Modules/socketmodule.c

```diff
@@ -6588,28 +6590,54 @@ Set the default timeout in seconds (float) for new socket objects.\n\
 A value of None indicates that new socket objects have no timeout.\n\
 When the socket module is first imported, the default is None.");
 
-#ifdef HAVE_IF_NAMEINDEX
+#if defined(HAVE_IF_NAMEINDEX) || defined(MS_WINDOWS)
 /* Python API for getting interface indices and names */
 
 static PyObject *
 socket_if_nameindex(PyObject *self, PyObject *arg)
 {
-    PyObject *list;
+    PyObject *list = PyList_New(0);
+    if (list == NULL) {
+        return NULL;
+    }
+#ifdef MS_WINDOWS
+    PMIB_IF_TABLE2 tbl;
+    if (GetIfTable2Ex(MibIfTableRaw, &tbl) != NO_ERROR) {
+        Py_DECREF(list);
+        return PyErr_SetFromErrno(PyExc_OSError);
+    }
+    for (ULONG i = 0; i < tbl->NumEntries; ++i) {
+        MIB_IF_ROW2 r = tbl->Table[i];
+        WCHAR buf[NDIS_IF_MAX_STRING_SIZE + 1];
+        if (ConvertInterfaceLuidToNameW(&r.InterfaceLuid, buf,
+                                        NDIS_IF_MAX_STRING_SIZE + 1)) {
+            Py_DECREF(list);
+            FreeMibTable(tbl);
+            return PyErr_SetFromErrno(PyExc_OSError);
+        }
+        PyObject *tuple = Py_BuildValue("IN", r.InterfaceIndex,
+                PyUnicode_FromWideChar(buf, -1));
```

Comment:
`Py_BuildValue` can take `u` and a `wchar_t*` and will do the string conversion for you.

Reply:
Done.